lld/lld-c/LLDAsLibraryC.cpp

```diff
@@ -76,23 +163,51 @@ SECTIONS {                                                                  \n\
     *(.data)                                                                \n\
   }}");
 
-  return (scriptPart1 + padding + "\n" + metadata + "\n" + scriptPart2).str();
+  return (linkerSymbolsStr + scriptPart1 + padding + scriptPart2).str();
 }
 
 LLVMBool LLVMLinkEraVM(LLVMMemoryBufferRef inBuffer,
-                       LLVMMemoryBufferRef *outBuffer, const char *metadataPtr,
-                       uint64_t metadataSize, char **ErrorMessage) {
-  llvm::SmallVector<llvm::MemoryBufferRef> localInMemBufRefs(2);
-  localInMemBufRefs[0] = *llvm::unwrap(inBuffer);
+                       LLVMMemoryBufferRef *outBuffer,
+                       const char *const *linkerSymbols,
+                       const char linkerSymbolValues[][LINKER_SYMBOL_SIZE],
+                       uint64_t numLinkerSymbols, char **errorMessage) {
+  SmallVector<MemoryBufferRef> localInMemBufRefs(2);
+  localInMemBufRefs[0] = *unwrap(inBuffer);
 
-  std::string linkerScript = creteEraVMLinkerScript(metadataPtr, metadataSize);
+  std::unique_ptr<Binary> InBinary =
+      cantFail(createBinary(unwrap(inBuffer)->getMemBufferRef()));
+  assert(InBinary->isObject());
 
-  std::unique_ptr<llvm::MemoryBuffer> linkerScriptBuf =
-      llvm::MemoryBuffer::getMemBuffer(linkerScript, "1");
+  bool shouldEmitRelocatable =
+      hasUndefLinkerSymbols(*static_cast<ObjectFile *>(InBinary.get()),
+                            linkerSymbols, numLinkerSymbols);
+
+  // Input ELF has undefined linker symbols, but no one definition was
+  // provided, which means the linker has nothing to do with the ELF file,
+  // so just copy it to the out buffer.
+  if (shouldEmitRelocatable && !numLinkerSymbols) {
+    StringRef inData = localInMemBufRefs[0].getBuffer();
+    *outBuffer = LLVMCreateMemoryBufferWithMemoryRangeCopy(
+        inData.data(), inData.size(), "result");
+    return false;
+  }
+
+  uint64_t MDSize = getSectionSize(*static_cast<ObjectFile *>(InBinary.get()),
+                                   ".eravm-metadata");
+
+  std::string linkerScript =
```

Comment:
LinkerSymbolDefs isn't a linker script, is it?
Perhaps worth refactoring or documenting what it means and implies if
* we don't have undef symbols
* we have undefs but 0 linkerSymbols
* we have have both undefs and linker symbols.

Otherwise, the logic below seems convoluted.

Reply:
In both cases we generate linker script. In case we are going emit an ELF file, the script is trivial, as it just contains linker symbol definitions. I've added the comments.